### PR TITLE
Support tf.function when model has ElasticDL embedding layer

### LINKED ISSUE #1188: Support tf.function when model has ElasticDL embedding layer
Currently, ElasticDL embedding layer looks up embedding vectors from Redis and use `tf.GradientTape.watch(embedding_vectors)` to tell gradient tape to record gradients of embedding vectors. However, in a function decorated by `tf.function`, `tape.watch` does not work.

We can use trainable variables instead. In a function decorated by `tf.function`, TensorFlow supports assigning values to variables, and gradient tape can automatically record gradients of trainable variables. So, we can create a TensorFlow variable called `batch_embedding` in advance. Every time worker does forward-pass, embedding layer assigns the embedding vectors to the `batch_embedding`.

One detail is worthy mention. Currently, ElasticDL supports the situation that an embedding layer is called more than once in one forward-pass. Please refer to this model as an example:

https://github.com/sql-machine-learning/elasticdl/blob/db0a1de92f3550878745feaea90c2f9fa0a2454b/elasticdl/python/tests/embedding_test_module.py#L7-L41

However, it is not simple to support this feature under `tf.function` because we plan to create TensorFlow variable in advance. We don't think this is an important feature. Thus we plan to not use `tf.function` if an embedding layer is called more than once.

### PR DESCRIPTION
**Pull request by mhaoli**
Fixes #1188.

This PR includes following:
* As #1188 states, create trainable variables in ElasticDL embedding layer.
* In `layer_test.py`, existing test cases only test embedding layer in eager mode. Modifying these test cases to test embedding layer both in eager mode and in graph mode (i.e. without `tf.function` and with `tf.function`).
* In `worker_test.py`, add a test case to test that embedding layer behaves the same when runs with  `tf.function` and without `tf.function`.

### CODE REVIEW

File: elasticdl/python/elasticdl/layers/embedding.py [hunk 1]
```diff
@@ -148,19 +172,27 @@ def call(self, input):
...
+                self.tape.watch(batch_embedding)
+                self._bet_ids_pair_eagerly.append((batch_embedding, flat_ids))
+                batch_embedding = self._bet_ids_pair_eagerly[-1][0]
+            else:
+                # In graph mode, assigning tensor to trainable variables is
+                # allowed and tape can record the gradients of trainable
+                # variables automatically.
+                self._bet_ids_pair_graph[0][0].assign(batch_embedding)
+                self._bet_ids_pair_graph[0][1].assign(flat_ids)
+                batch_embedding = self._bet_ids_pair_graph[0][0]
```
Comment (terrytangyuan) [inline elasticdl/python/elasticdl/layers/embedding.py]: It's a little hard to read these indices in the above code, e.g. `[-1][0]`. Consider assigning them as variables so they are more meaningful.

File: elasticdl/python/elasticdl/layers/embedding.py [hunk 2]
```diff
@@ -176,27 +208,43 @@ def _sparse_input_call(self, sparse_input):
...
+                    (batch_embedding, unique_ids)
+                )
+                batch_embedding = self._bet_ids_pair_eagerly[-1][0]
+            else:
+                # In graph mode, assigning tensor to trainable variables is
+                # allowed and tape can record the gradients of trainable
+                # variables automatically.
+                self._bet_ids_pair_graph[0][0].assign(batch_embedding)
+                self._bet_ids_pair_graph[0][1].assign(unique_ids)
+                batch_embedding = self._bet_ids_pair_graph[0][0]
```
Comment (terrytangyuan) [inline elasticdl/python/elasticdl/layers/embedding.py]: Repeated code?

File: elasticdl/python/tests/example_test.py [hunk 3]
```diff
@@ -124,7 +124,7 @@ def distributed_train_and_evaluate(
         )
 
         if dataset == "imagenet":
-            shards = {create_imagenet_recordio_file(16, feature_shape): 16}
+            shards = {create_imagenet_recordio_file(8, feature_shape): 8}
```
Comment (terrytangyuan) [inline elasticdl/python/tests/example_test.py:128]: Why this change?

File: elasticdl/python/tests/layer_test.py [hunk 4]
```diff
@@ -145,11 +142,19 @@ def test_embedding_layer(self):
...
+                lambda: module_call(model, inputs),
+            ]
+            for call_fn in call_fns:
+                outputs = call_fn()
+                values = outputs.numpy()[1]
+                for index, idx in enumerate(ids):
+                    correct_value = np.array(
+                        [idx] * output_dim, dtype=np.float32
+                    )
+                    self.assertTrue((values[index] == correct_value).all())
```
Comment (terrytangyuan) [inline elasticdl/python/tests/layer_test.py]: Repeated code?

File: elasticdl/python/tests/layer_test.py [hunk 5]
```diff
@@ -160,11 +165,13 @@ def test_embedding_layer_with_input_length(self):
...
-        for index, idx in enumerate(flatten_ids):
-            correct_value = np.array([idx] * output_dim, dtype=np.float32)
-            self.assertTrue((values[index] == correct_value).all())
+        call_fns = [lambda: layer.call(ids), lambda: module_call(layer, ids)]
+        for call_fn in call_fns:
+            values = call_fn()
+            values = values.numpy().reshape(-1, output_dim)
+            for index, idx in enumerate(flatten_ids):
+                correct_value = np.array([idx] * output_dim, dtype=np.float32)
+                self.assertTrue((values[index] == correct_value).all())
```
Comment (terrytangyuan) [inline elasticdl/python/tests/layer_test.py]: Same here and also other places (double check though)

File: elasticdl/python/tests/example_test.py (hunk 3 above)
Comment (mhaoli) [reply elasticdl/python/tests/example_test.py:128]: Because  the default memory limit of docker is 2G. Running ResNet50 with batch size 16 uses nearly 2G memory and sometimes it fails because of OOM.

File: elasticdl/python/elasticdl/layers/embedding.py (hunk 1 above)
Comment (mhaoli) [reply elasticdl/python/elasticdl/layers/embedding.py]: How about using named tuple?

Comment (terrytangyuan) [reply elasticdl/python/elasticdl/layers/embedding.py]: Sounds good to me

File: elasticdl/python/tests/example_test.py (hunk 3 above)
Comment (terrytangyuan) [reply elasticdl/python/tests/example_test.py:128]: Got it.

**Review by skydoorkai — APPROVED** [review]

File: elasticdl/python/elasticdl/layers/embedding.py (hunk 2 above)
Comment (mhaoli) [reply elasticdl/python/elasticdl/layers/embedding.py]: Done.

File: elasticdl/python/tests/layer_test.py (hunk 5 above)
Comment (mhaoli) [reply elasticdl/python/tests/layer_test.py]: Remove most of repeat code. Leave two test cases unchanged. Because removing a small quantity of repeat code in these two function requires changing the logic of test case itself. (`test_embedding_layer_gradient` and `test_embedding_layer_gradient_with_sparse_inputs`)
